### PR TITLE
move relations state server side

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

Move the uniter's internal relation state from files on the unit to the controller.

The uniter's internal relation state keeps track of relations made for this unit.  The entire relation.State structure will now be saved, rather than pieces and complicated ways of interpreting files to come with a relation.State.  

Upgrade steps included.  Migration steps for all of the uniter internal state moving server side will be handled together.

## QA steps
```console
$ juju bootstrap
$ juju deploy keystone 
$ juju deploy percona-cluster mysql 
$ juju add-relation keystone:shared-db mysql:shared-db 

# once the units stable, check the juju db
juju-db.bash
juju:PRIMARY> db.unitstates.find({},{"relation-state":1}).pretty()

# add a few units and relate them:
$ juju deploy glance -n 2 ; juju add-relation glance keystone  ; juju add-relation glance mysql

# once the units stable, check the juju db as above

# remove a glance unit
$ juju remove-unit glance/1

# once the units stable, check the juju db as above
```

Run same config upgrading from 2.7.5 to develop

### CODE REVIEW

**Comment by hmlanigan** [comment]
I need to fix the worker/uniter tests.  But the rest should be good to start review.

**Comment by hmlanigan** [comment]
Hoping the worker/uniter tests are now fixed.

File: worker/uniter/uniter.go [hunk 1]
```diff
@@ -683,6 +679,10 @@ func (u *Uniter) init(unitTag names.UnitTag) (err error) {
 	}
 	u.operationExecutor = operationExecutor
 
+	// Ensure we have an agent directory to to write the socket.
+	if err := os.MkdirAll(u.paths.State.BaseDir, 0755); err != nil {
```
Comment (hmlanigan) [inline worker/uniter/uniter.go:683]: We may be able to move this to the uniter tests, as in the real world it appears that it's not needed.  More research necessary.  Will investigate as a refinement.

File: upgrades/steps_28.go [hunk 2]
```diff
@@ -321,6 +331,137 @@ func readStorageStateFile(dirPath string, tag names.StorageTag) (bool, error) {
...
+			return relationData{}, errors.Annotatef(err, "failed to unmarshal relation state from %q", relationStateDir)
+		}
+		retData.data[k] = string(dataYaml)
+	}
+	return retData, nil
+}
+
+// readAllRelationStateDirs loads and returns every relationState persisted directly inside
+// the supplied dirPath. If dirPath does not exist, no error is returned.
+func readAllRelationStateDirs(dirPath string) (dirs map[int]*relationState, err error) {
```
Comment (hmlanigan) [inline upgrades/steps_28.go:364]: this was taken from worker/uniter/relation/state.go

File: upgrades/steps_28.go [hunk 3]
```diff
@@ -321,6 +331,137 @@ func readStorageStateFile(dirPath string, tag names.StorageTag) (bool, error) {
...
+		}
+		dirs[relationId] = rState
+	}
+	return dirs, nil
+}
+
+// readRelationStateDir loads a relationState from the subdirectory of dirPath named
+// for the supplied RelationId. If the directory does not exist, no error
+// is returned,
+func readRelationStateDir(dirPath string, relationId int) (d *relationState, err error) {
```
Comment (hmlanigan) [inline upgrades/steps_28.go:396]: this was taken from worker/uniter/relation/state.go

File: state/unitstate.go [hunk 4]
```diff
@@ -78,12 +78,14 @@ func (d *unitStateDoc) relationData() (map[int]string, error) {
...
 // relationStateMatches returns true if the RelationState map within the
-// unitStateDoc matches the provided newRS argument.  Assumes that
-// IgnoreRelationsState has been called first, therefore if arg is empty,
-// returns a nil map to be set.
+// unitStateDoc contains all of the provided newRS map.  Assumes that
+// relationStateBSONFriendly has been called first.
 func (d *unitStateDoc) relationStateMatches(newRS map[string]string) bool {
-	for k, v := range d.RelationState {
-		if newRS[k] != v {
+	if len(d.RelationState) < 1 {
```
[the remark this answers is not in the document]
Comment (achilleasa) [reply state/unitstate.go:84]: Without the proposed change, the for loop below would not detect changes if:
- newRS is a subset of oldRS
- the intersection of the two sets is not the empty set (iow, oldRS contains keys not present in newRS)

File: upgrades/steps_28.go [hunk 5]
```diff
@@ -200,6 +202,14 @@ func iaasUniterState(ctx Context, unitNames []string) ([]string, error) {
 			fileNames.Add(storageData.filename)
 		}
 
+		relationData, err := readRelationState(uniterStateDir)
```
Comment (achilleasa) [inline upgrades/steps_28.go:205]: Please remove TODO placeholder on L193

File: upgrades/steps_28.go [hunk 6]
```diff
@@ -200,6 +202,14 @@ func iaasUniterState(ctx Context, unitNames []string) ([]string, error) {
...
+		if err != nil && !os.IsNotExist(err) && !errors.IsNotFound(err) {
```
Comment (achilleasa) [inline upgrades/steps_28.go:206]: Since we control `readRelationState` why not have it test for `os.IsNotExist` internally and return back errors.IsNotFound? An example of this pattern can be found [here](https://github.com/juju/juju/blob/develop/worker/meterstatus/state.go#L112).

If you do this, can you also please change `readStorageState` above to use the same pattern?

File: worker/uniter/relation/relationer.go [hunk 7]
```diff
@@ -53,15 +53,15 @@ func (r *Relationer) RelationUnit() *apiuniter.RelationUnit {
 
 // Join initializes local state and causes the unit to enter its relation
 // scope, allowing its counterpart units to detect its presence and settings
-// changes. Local state directory is not created until needed.
+// changes.
 func (r *Relationer) Join() error {
 	if r.dying {
 		panic("dying relationer must not join!")
```
Comment (achilleasa) [inline worker/uniter/relation/relationer.go:59]: As a drive-by, can you replace this panic with an error?

File: worker/uniter/relation/relationer.go [hunk 8]
```diff
@@ -53,15 +53,15 @@ func (r *Relationer) RelationUnit() *apiuniter.RelationUnit {
...
 	}
-	// We need to make sure the state directory exists before we join the
-	// relation, lest a subsequent ReadAllStateDirs report local state that
+	// We need to make sure the state is persisted inState before we join
+	// the relation, lest a subsequent ReadAllStates report local state that
 	// doesn't include relations recorded in remote state.
-	if err := r.dir.Ensure(); err != nil {
+	if err := r.ops.EnsureState(); err != nil {
```
Comment (achilleasa) [inline worker/uniter/relation/relationer.go:64]: Took my a while to figure out what we mean by ensure; since we are refactoring could we rename it to something like `RefreshState` or `ReloadState` that will make sense to us when we revisit the code a few months down the road?

File: worker/uniter/relation/relationer_test.go [hunk 9]
```diff
@@ -304,23 +280,29 @@ func (s *RelationerImplicitSuite) TestImplicitRelationer(c *gc.C) {
...
 	c.Assert(err, jc.ErrorIsNil)
 
-	r := relation.NewRelationer(apiRelUnit, dir)
+	r := relation.NewRelationer(apiRelUnit, ops)
 	c.Assert(r, jc.Satisfies, (*relation.Relationer).IsImplicit)
 
 	// Hooks are not allowed.
-	f := func() { r.PrepareHook(hook.Info{}) }
+	f := func() { _, _ = r.PrepareHook(hook.Info{}) }
 	c.Assert(f, gc.PanicMatches, "implicit relations must not run hooks")
```
Comment (achilleasa) [inline worker/uniter/relation/relationer_test.go:288]: It would be a good opportunity to get rid of panics. If we bubble up the error, only the uniter worker will shutdown (and restart).

File: worker/uniter/relation/resolver.go [hunk 10]
```diff
@@ -153,21 +153,21 @@ func (r *relationsResolver) nextHookForRelation(localStateDir *StateDir, remote
...
+	for unitName := range localState.Members {
 		allUnitNames.Add(unitName)
 	}
 	for unitName := range remote.Members {
 		allUnitNames.Add(unitName)
 	}
 	sortedUnitNames := allUnitNames.SortedValues()
 	if allUnitNames.Contains("") {
-		return hook.Info{}, errors.Errorf("somehow we got the empty unit. local: %v, remote: %v", local.Members, remote.Members)
+		return hook.Info{}, errors.Errorf("somehow we got the empty unit. localState: %v, remote: %v", localState.Members, remote.Members)
```
Comment (achilleasa) [inline worker/uniter/relation/resolver.go:164]: small nit: s/remote: %v/remoteState: %v/?

File: worker/uniter/relation/resolver_test.go [hunk 11]
```diff
@@ -142,23 +140,26 @@ func (s *relationResolverSuite) setupRelations(c *gc.C) relation.RelationStateTr
...
-			CharmDir:             s.stateDir,
-			RelationsDir:         s.relationsDir,
+			Unit:                 u,
+			CharmDir:             s.charmDir,
 			NewLeadershipContext: s.leadershipContextFunc,
 			Abort:                abort,
 		})
 	c.Assert(err, jc.ErrorIsNil)
-	assertNumCalls(c, &numCalls, 3)
+	assertNumCalls(c, &numCalls, 4)
```
Comment (achilleasa) [inline worker/uniter/relation/resolver_test.go:162]: I had to resort to the same whack-a-mole game when I did the previous refactoring work here. We should really really try to rewrite these tests at some point to use mocks in the correct way...

File: worker/uniter/relation/resolver_test.go [hunk 12]
```diff
@@ -634,19 +720,13 @@ func (s *relationResolverSuite) TestHookRelationBrokenWhenSuspended(c *gc.C) {
 	relationsResolver := relation.NewRelationResolver(r, nil)
 	op, err := relationsResolver.NextOp(localState, remoteState, &mockOperations{})
 	c.Assert(err, jc.ErrorIsNil)
-	assertNumCalls(c, &numCalls, 11)
+	assertNumCalls(c, &numCalls, 18)
```
Comment (achilleasa) [inline worker/uniter/relation/resolver_test.go:723]: That's a rather significant bump in the number of API calls. Should we be worried?

File: worker/uniter/relation/resolver_test.go [hunk 13]
```diff
@@ -1018,13 +1158,24 @@ func principalWithSubordinateAPICalls() []apiCall {
...
+		}}
+	unitStateResults := params.UnitStateResults{Results: []params.UnitStateResult{{}}}
+
 	return []apiCall{
 		uniterAPICall("Refresh", nrpeUnitEntity, params.UnitRefreshResults{Results: []params.UnitRefreshResult{{Life: life.Alive, Resolved: params.ResolvedNone}}}, nil),
 		uniterAPICall("GetPrincipal", nrpeUnitEntity, params.StringBoolResults{Results: []params.StringBoolResult{{Result: "unit-wordpress-0", Ok: true}}}, nil),
 		uniterAPICall("RelationsStatus", nrpeUnitEntity, relationStatusResults, nil),
 		uniterAPICall("Relation", relationUnits1, relationResults1, nil),
+		uniterAPICall("State", nrpeUnitEntity, unitStateResults, nil),
+		//uniterAPICall("State", nrpeUnitEntity, unitStateResults, nil),
```
Comment (achilleasa) [inline worker/uniter/relation/resolver_test.go:1175]: Can the commented line be removed?

File: worker/uniter/relation/state.go [hunk 14]
```diff
@@ -108,200 +107,196 @@ func (s *State) Validate(hi hook.Info) (err error) {
...
+type UnitStateReadWriter interface {
+	State() (params.UnitStateResult, error)
+	SetState(unitState params.SetUnitStateArg) error
 }
 
-// State returns the current state of the relation.
-func (d *StateDir) State() *State {
-	return d.state.copy()
+// StateOps is a representation of the relation state. Concurrent
+// modifications to the underlying state will have undefined consequences.
```
Comment (achilleasa) [inline worker/uniter/relation/state.go:118]: Nice to include concurrency limitations in docs for the type!

File: worker/uniter/relation/state.go [hunk 15]
```diff
@@ -108,200 +107,196 @@ func (s *State) Validate(hi hook.Info) (err error) {
...
-func (d *StateDir) Write(hi hook.Info) (err error) {
-	defer errors.DeferredAnnotatef(&err, "failed to write %q hook info for %q on state directory", hi.Kind, hi.RemoteUnit)
+func (s *StateOps) WriteHook(hi hook.Info) (err error) {
+	defer errors.DeferredAnnotatef(&err, "failed to write %q hook info for %q in state", hi.Kind, hi.RemoteUnit)
 	if hi.Kind == hooks.RelationBroken {
-		return d.Remove()
+		return s.Remove()
 	}
-	name := strings.Replace(hi.RemoteUnit, "/", "-", 1)
 	isApp := false
```
Comment (achilleasa) [inline worker/uniter/relation/state.go:193]: Due to the removal, this block can be condensed to `isApp := hi.RemoteUnit == ""`

File: worker/uniter/relation/state.go [hunk 16]
```diff
@@ -108,200 +107,196 @@ func (s *State) Validate(hi hook.Info) (err error) {
...
 	//  delete "foo-app". Instead, during relation-broken, we cleanup all related applications.
-	for appMember := range d.state.ApplicationMembers {
-		path := filepath.Join(d.path, appMember+"-app")
-		if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
-			return errors.Trace(err)
-		}
+	//
+	// Note(hml): 2020-04-01 We no longer keep persisted state as local files,
+	// instead remove all ApplicationMembers.
+	s.state.ApplicationMembers = nil
```
Comment (achilleasa) [inline worker/uniter/relation/state.go:249]: I believe this should be at the bottom of the method after L269; we don't want to reset this map before the `SetState` call returns successfully

File: worker/uniter/relation/state.go [hunk 17]
```diff
@@ -108,200 +107,196 @@ func (s *State) Validate(hi hook.Info) (err error) {
...
-	ChangeVersion  *int64 `yaml:"change-version"`
-	ChangedPending bool   `yaml:"changed-pending,omitempty"`
+// State returns the current state of the relation.
+func (s *StateOps) State() *State {
+	return s.state.copy()
+}
+
+// EnsureState writes state via SetState as ReadState returns a
+// new state even if one was not read from State.
+func (s *StateOps) EnsureState() error {
```
Comment (achilleasa) [inline worker/uniter/relation/state.go:280]: As mentioned above; EnsureState is a bit misleading method name.

File: worker/uniter/relation/state.go [hunk 18]
```diff
@@ -108,200 +107,196 @@ func (s *State) Validate(hi hook.Info) (err error) {
...
+	State() (params.UnitStateResult, error)
+	SetState(unitState params.SetUnitStateArg) error
 }
 
-// State returns the current state of the relation.
-func (d *StateDir) State() *State {
-	return d.state.copy()
+// StateOps is a representation of the relation state. Concurrent
+// modifications to the underlying state will have undefined consequences.
+type StateOps struct {
```
Comment (achilleasa) [inline worker/uniter/relation/state.go:119]: Is there a particular reason why we don't use a single object to load the complete state data at construction time (just like ReadAllStates) and then provide an API to read/write the data for a particular relation ID? This could be something created by the uniter and passed through to the relation state tracker and from there to the relationers.

This way, your data will always be up to date (and if the uniter restarts it will pull a fresh copy) and you don't need to have calls like `Ensure` in there.

Comment (hmlanigan) [reply worker/uniter/relation/state.go]: StateOps have now been replaced with StateManager which handles all of the relation state.

File: state/unit_ops.go [hunk 19]
```diff
@@ -130,6 +130,10 @@ func (op *unitSetStateOperation) newUnitStateDoc(unitGlobalKey string) (unitStat
 		newStDoc.StorageState = storState
 		quotaChecker.Check(storState)
 	}
+	if relState, found := op.newState.relationStateBSONFriendly(); found {
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply state/unit_ops.go]: didn't see the existing code, read too fast given the order was different.  removing

File: state/unitstate.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hmlanigan) [reply state/unitstate.go]: done

File: upgrades/steps_28.go (hunk 5 above)
Comment (hmlanigan) [reply upgrades/steps_28.go:193]: done, removed another todo as well.

File: upgrades/steps_28.go (hunk 6 above)
Comment (hmlanigan) [reply upgrades/steps_28.go]: for the relation piece  IsNotFound will not occur.  it's specifically returned in the readStorageState though.  I can replace the NotExists errors with NotFound for storage.

File: worker/uniter/relation/state.go (hunk 15 above)
Comment (hmlanigan) [reply worker/uniter/relation/state.go]: done

File: worker/uniter/relation/relationer.go (hunk 7 above)
Comment (hmlanigan) [reply worker/uniter/relation/relationer.go:65]: I'd rather not do this as a drive-by.  Will double check that the uniter will be restarted on an error here and address in a new PR

File: upgrades/steps_28.go [hunk 20]
```diff
@@ -200,6 +190,14 @@ func iaasUniterState(ctx Context, unitNames []string) ([]string, error) {
 			fileNames.Add(storageData.filename)
 		}
 
+		relationData, err := readRelationState(uniterStateDir)
+		if err != nil && !os.IsNotExist(err) {
```
Comment (achilleasa) [inline upgrades/steps_28.go:194]: Just to double-check: is os.IsNotExist the right error type to check here? We don't seem to wrap it in a not found error, right?

File: upgrades/steps_28.go [hunk 21]
```diff
@@ -321,6 +325,137 @@ func readStorageStateFile(dirPath string, tag names.StorageTag) (bool, error) {
...
+	}
+	return retData, nil
+}
+
+// readAllRelationStateDirs loads and returns every relationState persisted directly inside
+// the supplied dirPath. If dirPath does not exist, no error is returned.
+func readAllRelationStateDirs(dirPath string) (dirs map[int]*relationState, err error) {
+	defer errors.DeferredAnnotatef(&err, "cannot load relations state from %q", dirPath)
+	if _, err := os.Stat(dirPath); os.IsNotExist(err) {
+		return nil, nil
```
Comment (achilleasa) [inline upgrades/steps_28.go:361]: If the dir does not exist, we return nil, and `readRelationState` returns back an empty map. Is this the intended behavior?

File: upgrades/steps_28.go (hunk 20 above)
Comment (achilleasa) [reply upgrades/steps_28.go:194]: It looks like `readRelationState` and the functions it calls never return errors that pass the `os.IsNotExist` predicate.

File: upgrades/steps_28.go [hunk 22]
```diff
@@ -321,6 +325,137 @@ func readStorageStateFile(dirPath string, tag names.StorageTag) (bool, error) {
...
+func readRelationStateDir(dirPath string, relationId int) (d *relationState, err error) {
+	d = &relationState{
+		RelationId:         relationId,
+		Members:            map[string]int64{},
+		ApplicationMembers: map[string]int64{},
+		ChangedPending:     "",
+	}
+	defer errors.DeferredAnnotatef(&err, "cannot load relation state from %q", dirPath)
+	if _, err := os.Stat(dirPath); os.IsNotExist(err) {
+		return d, nil
```
Comment (achilleasa) [inline upgrades/steps_28.go:399]: Same question about suppressing the error and returning back a blank relationState. So, if we find an _empty_ folder for the relation we silently convert it into a blank state? Wouldn't the absense of state data indicate that something is wrong?

File: worker/uniter/relation/relationer.go (hunk 7 above)
Comment (achilleasa) [reply worker/uniter/relation/relationer.go:65]: Fair enough; we can definitely do this in a separate PR as you suggest.

File: worker/uniter/relation/statemanager.go [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+// UnitStateReadWriter encapsulates the methods from a state.Unit
+// required to set and get unit state.
+type UnitStateReadWriter interface {
+	State() (params.UnitStateResult, error)
+	SetState(unitState params.SetUnitStateArg) error
+}
+
+// NewStateManager
```
Comment (achilleasa) [inline worker/uniter/relation/statemanager.go:46]: nit: Missing comment ;-)

File: worker/uniter/relation/statemanager.go [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+type stateManager struct {
+	unitStateRW   UnitStateReadWriter
+	relationState map[int]State
+	mu            sync.Mutex
+}
+
+// RelationState returns a copy of the relation state for
+// the given id. Returns NotFound.
```
Comment (achilleasa) [inline worker/uniter/relation/statemanager.go:59]: Slight correction;
s/. Returns NotFound/ or returns NotFound if the given id does not belong to a known relation/

File: worker/uniter/relation/relationer_test.go [hunk 25]
```diff
@@ -4,43 +4,42 @@
 package relation_test
 
 import (
-	"strconv"
+	"github.com/juju/juju/apiserver/params"
```
Comment (hmlanigan) [inline worker/uniter/relation/relationer_test.go:7]: I need to fix the stanzas

File: upgrades/steps_28.go (hunk 20 above)
Comment (hmlanigan) [reply upgrades/steps_28.go:194]: It can bubble up from os.Stat in readAllRelationStateDirs.  Although, I wonder if we shouldn't fail an upgrade on IsNotExist because there should always be a peer relation to move. readAllRelationStateDirs was moved from the uniter and dealt with uniter start up scenarios too.

**Comment by achilleasa** [comment]
QA works as expected!

File: upgrades/steps_28_test.go [hunk 26]
```diff
@@ -117,15 +117,16 @@ func (s *mockSteps28Suite) SetUpTest(c *gc.C) {
...
-
 	s.opStorOneYaml, s.opStorOneFileName = writeStorageState(c, unitOneStateDir, s.storTagOne, s.opStorOne)
-
 	s.opRelationYaml, s.opRelationFileName = setupRelationState(c, unitOneStateDir)
 
 	unitTwoStateDir := filepath.Join(agentDir, s.tagTwo.String(), "state")
 	err = os.MkdirAll(unitTwoStateDir, 0755)
 	c.Assert(err, jc.ErrorIsNil)
-	s.opStateTwoYaml, s.opStateTwoFileName = writeUnitStateFile(c, unitTwoStateDir, s.opStateTwo)
+	//s.opStateTwoYaml, s.opStateTwoFileName = writeUnitStateFile(c, unitTwoStateDir, s.opStateTwo)
```
Comment (achilleasa) [inline upgrades/steps_28_test.go:126]: Don't forget to remove this before merging.

**Comment by hmlanigan** [comment]
rebased

**Comment by hmlanigan** [comment]
$$merge$$

**Comment by hmlanigan** [comment]
$$merge$$
